Feat: Add Skeleton to support Dataset anonymization with differential privacy (#455)
cmu-db/benchbase: scripts/anonymization/README.md

Reviewer: Please add a short section on setup/requirements (e.g., python 3.10, pip install requirements.txt, etc.) and running tests locally (e.g., pytest)

# Anonymization

> <span style="color:orange"> ANONYMIZATION IS A WORK IN PROGRESS AND DOES CURRENTLY NOT ACTUALLY ANONYMIZE THE DATA. THIS FEATURE WILL BE ADDED LATER </span>

Reviewer: ```suggestion
*NOTE: ANONYMIZATION IS A WORK IN PROGRESS AND DOES CURRENTLY NOT ACTUALLY ANONYMIZE THE DATA. THIS FEATURE WILL BE ADDED LATER*
```

> The anonymization module allows to apply privacy mechanisms such as differential privacy or column faking to the data.

Reviewer: ```suggestion
The anonymization module allows applying privacy mechanisms such as differential privacy or column faking to the data.
```

Reviewer: Maybe also add links to external reference sources on differential privacy.

| ----------- | ------- | ----------------------------------- | -------------------------------------------------------------------------------------------- |
| name        | -       | Any string                          | The name of the table                                                                        |
| epsilon     | 1.0     | Any value > 0                       | The privacy budget. Higher values will decrease the privacy guarantee                        |
| pre_epsilon | 0.5     | Any value between 0 and epsilon     | The privacy budget spent on preprocessing of the data. Only necessary for continuous columns |
| algorithm   | mst     | One of: `mst,aim,dpctgan,patectgan` | The differential privacy mechanism applied to the data                                       |

```xml
<anonymization>
    <table
        name="tablename"
        epsilon="1.5"
        pre_epsilon="0.75"
        algorithm="patectgan"
    />
</anonymization>
```

### Column Information

The columns of a table can be split into four categories:

> 1. `drop` - Columns that are omitted from the DP-mechanism

Reviewer: Clarifying: 
`drop` -> "omitted" or "removed"?

i.e., are those columns skipped when applying differential privacy but still exist in the dataset at the end as they originally appeared,
or are the "removed" from the dataset (i.e, `ALTER TABLE DROP COLUMN ...`)

Reviewer: If the former, maybe call it something else like "passthru" to avoid confusion?

Author: renamed to "ignore". Description is in the README. Ignored columns are not anonymized but added back in their original form to the DB

4. `ordinal` - Columns of ordinal meaning. Typically numerical columns with low amount of distinct values

The names of the columns must be given as text. Multiple column names must be separated by comma without additional whitespace.

The anonymization process will automatically try to infer the category best suitable category of each column. Sometimes this leads to undesired results or is unable to process the column because of other factors. To counter this, each column can be defined by hand with the suitable tags.

---

**Drop the ID-column and anonymize automatically:**

```xml
<anonymization>
    <table name="item">
        <drop>i_id</drop>
    </table>
</anonymization>
```

The ID-column will be removed from the anonymization process and be added back to the data in its original form! This is useful for primary key columns or columns that contain only NULL values.

---

**Define the type of each column individually:**

```xml
<anonymization>
    <table name="item">
        <categorical>i_name,i_data,i_im_id</categorical>
        <continuous>i_price</continuous>
        <ordinal>i_id</ordinal>
    </table>
</anonymization>

> ```

Reviewer: Alternative XML schema:

```xml
<anonymization>
  <table name="t">
    <columns>
      <column name="c1" type="continuous" />
      <column name="c2" type="continuous" />
      <column name="c2" type="ordinal" />
      <column name="c3" type="categorical" />
      <column name="c4" type="passthru" />
  </table>
```

Reviewer: Basically, avoid sub parsing in another language (csv) and just use xml for the whole thing in some way.

Reviewer: Also, then all your "fine tuning" configs below can get merged into this rather than having two separate sections.

Author: Changed Schema to avoid sub parsing. See README

<continuousConfig>
```

The default parameters must be added inside of the `<column>`-tag. (See below)

| Name  | Default | Possible Value    | Description                                 |
| ----- | ------- | ----------------- | ------------------------------------------- |
| name  | -       | Any string        | The name of the column                      |
| bins  | 10      | Any value         | The amount of distinct values in the output |
| lower | -       | Any value         | Lower bound of the values                   |
| upper | -       | Any value > lower | Upper bound of the values                   |

---

A full working example could look like this:

```xml
<anonymization>
    <table name="item" epsilon="1.0" pre_epsilon="0.0" algorithm="aim">
        <!-- Column categorization -->
        <droppable>i_id</droppable>
        <categorical>i_name,i_data,i_im_id</categorical>
        <continuous>i_price</continuous>
        <!-- Continuous column fine-tuning -->
        <continuousConfig>
            <column name="i_price" bins="1000" lower="2.0" upper="100.0">
        <continuousConfig>
    </table>
</anonymization>
```

### Sensitive Values

> The anonymization module supports value faking to handle sensitive values. A `<sensitive>`-tag must be created to signify that column faking should be applied. Column faking can be defined individually on each column. To do so, parameters can be added inside of the `<column>`-tag. (See below)

Reviewer: ```suggestion
The anonymization module supports value faking to handle sensitive values.
A `<sensitive>`-tag must be created to signify that column faking should be applied.
Column faking can be defined individually on each column.
To do so, parameters can be added inside of the `<column>`-tag. (See below)
```

nit: one sentence per line

| Name    | Default | Possible Value                                   | Description                                                                                    |
| ------- | ------- | ------------------------------------------------ | ---------------------------------------------------------------------------------------------- |
| name    | -       | Any string                                       | The name of the column                                                                         |
| method  | pystr   | Any method provided by the python faker library  | The faking method of the library. If the method is not found, a random string will be inserted |
| locales | -       | A string of supported locales separated by comma | The locale to produce localized values like chinese street names or english first names        |
| seed    | 0       | Any integer value                                | The privacy budget spent on preprocessing of the data. Only necessary for continuous columns   |

```xml
<anonymization>
    <table name="item" algorithm="aim">

>         <sensitive>

Reviewer: Is "column faking" mutually exclusive with "differential privacy"?

Or can the schema somehow be unified with the one I mentioned above?

Author: We would like to handle it separately as part of the Anonymization.

This allows for anonymization of sensitive columns without differential privacy (For example for the Twitter workload, user_profiles table)

or coupled anonymization (Run DP first and then handle the sensitive values for higher privacy guarantees)

            <column
                name="i_name"
                method="name"
                locales="en_US"
                seed="0">
            </column>
            <column name="i_data" method="pystr" />
        </sensitive>
    </table>
</anonymization>
```

## Examples

### Basic

> The most basic config will need only the name of the table and apply default values for the anonymization.

Reviewer: Document current defaults, or where to look them up?

Author: Changed this part of the README. Automatic anonymization will automatically try to categorize columns. 

Defaults are documented and I provided a link to the algorithm repository as additional info

| Name        | Default | Possible Value                                     | Description                                                                                  |
| ----------- | ------- | -------------------------------------------------- | -------------------------------------------------------------------------------------------- |
| epsilon     | 1.0     | Any value > 0                                      | The privacy budget. Higher values will decrease the privacy guarantee                        |
| pre_epsilon | 0.5     | Any value between 0 and epsilon (epsilon excluded) | The privacy budget spent on preprocessing of the data. Only necessary for continuous columns |
| algorithm   | mst     | One of: `mst,aim,dpctgan,patectgan`                | The differential privacy mechanism applied to the data                                       |

```xml
<anonymization>
    <table name="tablename">
        <differential_privacy
            epsilon="1.5"
            pre_epsilon="0.75"
            algorithm="patectgan">
        </differential_privacy>
    </table>
</anonymization>
```

### Column Information

The columns of a table can be split into four categories:

> 1. `ignore` - Columns that are ignored by the DP-mechanism. Left untouched

Reviewer: ```suggestion
1. `ignore` - Columns that are ignored by the DP-mechanism. Left untouched.
```

nit: consistency about ending sentences of list items in `.` periods.

---

**Ignore the ID-column and anonymize automatically:**

```xml
<anonymization>
    <table name="item">
        <differential_privacy>
            <ignore>
                <column name="i_id"/>
            </ignore>
            <!-- Other types are inferred automatically -->
        </differential_privacy>
    </table>
</anonymization>
```

The ID-column will be removed from the anonymization process and be added back to the data in its original form before pushing back to the DB! This is useful for primary key columns or columns that contain only NULL values.

---

**Define the type of each column individually:**

```xml
<anonymization>
    <table name="item">
        <differential_privacy>
            <categorical>
                <column name="i_name"/>
                <column name="i_data"/>
                <column name="i_im_id"/>
            </categorical>
            <continuous>
                <column name="i_price"/>
            </continuous>
            <ordinal>
                <column name="i_id"/>
            </ordinal>
        </differential_privacy>
    </table>
</anonymization>
```

> <span style="color:red"> **Disclaimer**: As soon as the column types are defined by hand, all columns must be defined. It is not possible to only specify some of the categorical columns and let the algorithms do the rest!</span>

Reviewer: Question about this: Couldn't you
1) start with inferred/discovered defaults
2) use any additional column details in the config to override those defaults
so that you wouldn't have to either specify no columns or specify all columns?

Author: Unfortunately, this is the architecture of the DP-algorithms library that we use. (They do not allow mixed configurations)
